internal/controllers/gateway/httproute_controller.go

```diff
@@ -658,43 +659,47 @@ func (r *HTTPRouteReconciler) setRouteConditionResolvedRefsCondition(
 	return parentStatuses, changed, nil
 }
 
-func (r *HTTPRouteReconciler) getHTTPRouteRuleReason(ctx context.Context, httpRoute gatewayapi.HTTPRoute) (gatewayapi.RouteConditionReason, error) {
+func (r *HTTPRouteReconciler) getHTTPRouteRuleReason(ctx context.Context, httpRoute gatewayapi.HTTPRoute) (reason gatewayapi.RouteConditionReason, msg string, err error) {
 	for _, rule := range httpRoute.Spec.Rules {
 		for _, backendRef := range rule.BackendRefs {
 			backendNamespace := httpRoute.Namespace
 			if backendRef.Namespace != nil && *backendRef.Namespace != "" {
 				backendNamespace = string(*backendRef.Namespace)
 			}
 
+			backendRefGK := string(*backendRef.Kind)
+			if gr := string(*backendRef.Group); gr != "" {
+				backendRefGK = gr + "/" + backendRefGK
+			}
 			// Check if the BackendRef GroupKind is supported
 			if !util.IsBackendRefGroupKindSupported(backendRef.Group, backendRef.Kind) {
-				return gatewayapi.RouteReasonInvalidKind, nil
+				return gatewayapi.RouteReasonInvalidKind, fmt.Sprintf("target of type %s is unexpected", backendRefGK), nil
 			}
 
 			// Check if all the objects referenced actually exist
 			// Only services are currently supported as BackendRef objects
 			service := &corev1.Service{}
-			err := r.Client.Get(ctx, k8stypes.NamespacedName{Namespace: backendNamespace, Name: string(backendRef.Name)}, service)
-			if err != nil {
+			serviceNN := k8stypes.NamespacedName{Namespace: backendNamespace, Name: string(backendRef.Name)}
+			if err := r.Client.Get(ctx, serviceNN, service); err != nil {
 				if !apierrors.IsNotFound(err) {
-					return "", err
+					return "", "", err
 				}
-				return gatewayapi.RouteReasonBackendNotFound, nil
+				return gatewayapi.RouteReasonBackendNotFound, fmt.Sprintf("target %s of type %s does not exist", serviceNN, backendRefGK), nil
 			}
 
 			// Check if the object referenced is in another namespace,
 			// and if there is grant for that reference
 			if httpRoute.Namespace != backendNamespace {
 				if !r.enableReferenceGrant {
-					return gatewayapi.RouteReasonRefNotPermitted, nil
+					return gatewayapi.RouteReasonRefNotPermitted, "Usage of ReferenceGrants has to be explicitly enabled", nil
```

Comment:
Might be misleading to some. AFAIK this requires a cli flag/env to enable. I'd mention that in the message. WDYT?

Reply:
Good suggestion, I made it more detailed in all cases when permission is not granted. PTAL